### PR TITLE
WIP polygonal/polyhedral transformations

### PR DESCRIPTION
**Pull request by nselliott**
Showing current state of the transformations.  2D is pretty much there, lots to do for 3D.

### CODE REVIEW

**Comment by xjrc** [comment]
Thanks for putting this together, @nselliott; I'll give all this code a review tomorrow afternoon.

**Comment by xjrc** [comment]
I've had a brief look over the code and I have a request to make the review a bit easier: would you mind giving a brief 1-2 sentence synopsis on each new function and its role in the AMR-polytopal conversion process? I've compiled a list of all of the new functions I could find below:

- `conduit::blueprint::mpi::mesh::to_polygonal(const Node&, Node&, const std::string&)`
- `conduit::blueprint::mpi::mesh::to_polyhedral(const Node&, Node&, const std::string&)`
- `conduit::blueprint::mpi::mesh::generate_partition(const Node&, Node&, MPI_Comm)`
- `conduit::blueprint::mesh::connectivity::make_element_2d(std::vector<int64_t>&, int64_t, int64_t)`
- `conduit::blueprint::mesh::connectivity::make_element_3d(PolyElemType&, int64_t, int64_t, int64_t, int64_t, std::map<int, std::vector<int64_t>>&, std::map<int, std::vector<int64_t>>&, std::map<int, std::vector<int64_t>>&)`
- `conduit::blueprint::mesh::connectivity::create_elements_2d(const Node&, int64_t, int64_t, int64_t, std::map<int, std::vector<int64_t>>&)`
- `conduit::blueprint::mesh::connectivity::create_elements_3d(const Node&, int64_t, int64_t, int64_t, int64_t, int64_t, int64_t, std::map<int, PolyElemType>&, std::map<int, std::vector<int64_t>>&, std::map<int, std::vector<int64_t>>&, std::map<int, std::vector<int64_t>>&)`
- `conduit::blueprint::mesh::connectivity::connect_elements_2d(const Node&, int64_t, int64_t, int64_t, int64_t, int64_t&, std::map<int, std::vector<int64_t>>&)`
- `conduit::blueprint::mesh::connectivity::connect_elements_3d(const Node&, int64_t, int64_t, int64_t, int64_t, int64_t, int64_t&, std::map<int, PolyElemType>&)`

**Comment by nselliott** [comment]
> I've had a brief look over the code and I have a request to make the review a bit easier: would you mind giving a brief 1-2 sentence synopsis on each new function and its role in the AMR-polytopal conversion process? I've compiled a list of all of the new functions I could find below:

Certainly.  I suggest you mostly focus on the 2D functions, since the 3D functions are incomplete and may end up completely different.
 
>     * `conduit::blueprint::mpi::mesh::to_polygonal(const Node&, Node&, const std::string&)`

This is the user API call for the 2D transformation from a structured mesh to a polygonal mesh.

>     * `conduit::blueprint::mpi::mesh::to_polyhedral(const Node&, Node&, const std::string&)`

Likewise for 3D transformations from structured to polyhedral.  (This is not complete.)

>     * `conduit::blueprint::mpi::mesh::generate_partition(const Node&, Node&, MPI_Comm)`

This is not directly part of the transformations.  It is intended to be part of generate_index, to automatically set up the domain-to-rank partition map for the new multi-domain index format.
 
>     * `conduit::blueprint::mesh::connectivity::make_element_2d(std::vector<int64_t>&, int64_t, int64_t)`

This creates a single 2D quadrilateral polygonal element based on information from the structured topology.  The vertices go into the vector to be later accumulated into the larger connectivity array.  If the element needs to have extra vertices added at a domain boundary, that happens elsewhere.

>     * `conduit::blueprint::mesh::connectivity::make_element_3d(PolyElemType&, int64_t, int64_t, int64_t, int64_t, std::map<int, std::vector<int64_t>>&, std::map<int, std::vector<int64_t>>&, std::map<int, std::vector<int64_t>>&)`

This creates a single polyhedral element for a hexahedron computed from the structured topology.  It also collects some metadata that is intended for figuring out connections of faces across domain boundaries, but this will probably be revamped.
 
>     * `conduit::blueprint::mesh::connectivity::create_elements_2d(const Node&, int64_t, int64_t, int64_t, std::map<int, std::vector<int64_t>>&)`

This creates the 2d elements that are specifically located on the boundaries of 2D domains where they touch finer domains.  It calls make_elements_2d to set them up as quadrilaterals and to prepare them to have extra vertices added.

>     * `conduit::blueprint::mesh::connectivity::create_elements_3d(const Node&, int64_t, int64_t, int64_t, int64_t, int64_t, int64_t, std::map<int, PolyElemType>&, std::map<int, std::vector<int64_t>>&, std::map<int, std::vector<int64_t>>&, std::map<int, std::vector<int64_t>>&)`

Same thing for 3D.

>     * `conduit::blueprint::mesh::connectivity::connect_elements_2d(const Node&, int64_t, int64_t, int64_t, int64_t, int64_t&, std::map<int, std::vector<int64_t>>&)

This figures out the connections between the coarse elements that touch the domain boundary and their adjacent fine elements.  It adds new vertices to the coarse elements at the appropriate locations.

> 
>     * `conduit::blueprint::mesh::connectivity::connect_elements_3d(const Node&, int64_t, int64_t, int64_t, int64_t, int64_t, int64_t&, std::map<int, PolyElemType>&)`

Same for 3D, but it needs to figure out both new faces and new vertices, and it doesn't work.

File: src/libs/blueprint/conduit_blueprint_mesh.hpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mesh.hpp]: Given their nature as helper functions for the greater AMR-to-polytopal process, I think it would be best for all of these functions to be excluded from external-facing headers (similar to [`ShapeType`](https://github.com/LLNL/conduit/blob/develop/src/libs/blueprint/conduit_blueprint_mesh.cpp#L174) and its associated functionality).

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mesh.cpp]: I feel like the interface for this function is a bit ambiguous. In particular, the contents of the out variable `elem` could be one of many things (a list of *just* this 2D element's vertices etc.), where they are in fact an accumulated list of potentially many 2D elements. In order to make this behavior clearer, could we rename this function to something like `append_element_2d`?

Comment (xjrc) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp]: This question extends to the `{i|j|k}faces` output arguments to the `make_element_3d` function.

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mesh.cpp]: A minor nitpick, but the indentation for the `{i|j|k}faces` arguments is slightly off.

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mesh.cpp]: I feel like there has to be some good way to abstract the behavior used for each dimension in this function into a general routine across dimensions (e.g. something like [`grid_id_to_ijk`](https://github.com/LLNL/conduit/blob/develop/src/libs/blueprint/conduit_blueprint_mesh.cpp#L313)).

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mesh.cpp]: The argument indentation is off here as well.

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mesh.cpp]: > This creates the 2d elements that are specifically located on the boundaries of 2D domains where they touch finer domains. It calls make_elements_2d to set them up as quadrilaterals and to prepare them to have extra vertices added.

I'm having difficulty mapping the above description onto the arguments presented for the `create_elements_2d` function. In my mind, it seems sufficient to pass just the structured AMR mesh window and a direction to generate all of the boundary elements for that window. I can square this circle by mapping `i_lo` and `j_lo` onto some sort of boundary anchor, but the existence of `iwidth` (which, to me, seems redundant relative to the information in `ref_win`) makes me question whether I understand this function at all.

If you could Noah, could you please explain how this function selects the boundary from within the given `ref_win` using `i_lo`, `j_lo`, and `iwidth`?

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mesh.cpp]: The use of the output variable `new_vertex` here seems highly suggestive that this whole set of functions should be contained within a struct/class like [`ShapeType`](https://github.com/LLNL/conduit/blob/develop/src/libs/blueprint/conduit_blueprint_mesh.cpp#L174), as the purpose of such variables and how they're initialized/used becomes too opaque otherwise.

Comment (xjrc) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp]: Also, I was tempted to ask how you were able to get away with this function being simpler than `connect_elements_2d` (a bit of a hopeful leading question toward simplifying `connect_elements_2d`), but the fact that this function doesn't work indicates that it may become more complex as it becomes more complete.

File: src/libs/blueprint/conduit_blueprint_mpi_mesh.cpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mpi_mesh.cpp]: Can't all of these cascading `if` statements be combined into a single large clause via `&&`? For example:

```
if(par_rank == selected_rank &&
        ::conduit::blueprint::mesh::is_multi_domain(mesh) &&
        mesh.child(0).has_child("state") &&
        mesh.child(0)["state"].has_child("domain_id")) {
    do_par_map.set_uint8(1);
}
```

I feel like this style more clearly signals the intent; with the cascading `if` statements, I'm always on the lookout for extra behavior within each individual clause.

File: src/libs/blueprint/conduit_blueprint_mpi_mesh.cpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mpi_mesh.cpp]: For my own benefit, this naming scheme appears to have been derived from #476 (and specifically, [this comment](https://github.com/LLNL/conduit/issues/476#issuecomment-662120834)).

File: src/libs/blueprint/conduit_blueprint_mpi_mesh.hpp
Comment (xjrc) [inline src/libs/blueprint/conduit_blueprint_mpi_mesh.hpp]: On the one hand, I realize that these functions are peculiar because they apply mesh-wide transformations and thus don't fit nicely within an embedded namespace (e.g. `conduit::blueprint::mpi::mesh::topology::xyz::to_polygonal`), but I feel like more clarification is necessary to properly scope the set of functions to which this can be applied (i.e. structured AMR meshes). As an initial suggestion to start a conversation, would something like `from_amr_to_poly{gonal|hedral}` be an improvement?

File: src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp
Comment (xjrc) [inline src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp]: As this example is a bit complicated, would it be possible to accompany it with an ASCII diagram in a code comment?

File: src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp
Comment (xjrc) [inline src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp]: It looks like this code was lifted from the previous example; would it be possible to abstract this into a helper function?

**Review by xjrc — CHANGES_REQUESTED** [review]
Sorry it took me so long to get to this @nselliott, but I was able to go through most of it. Most of my comments are questions and some software architecture/organization suggestions, but overall the functionality looks good. I think we should converge on the final placement of some of this functionality before we should merge, but I think that it's close.

File: src/libs/blueprint/conduit_blueprint_mesh.hpp
Comment (nselliott) [reply src/libs/blueprint/conduit_blueprint_mesh.hpp]: So just put these helpers completely in conduit_blueprint_mpi_mesh.cpp?  That is fine, as I will defer to you on these kinds of code organization matters.  I separated them because they have no MPI, but if their only use is to enable MPI-based functionality, it still makes sense.

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (nselliott) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp]: This always creates one element defined by four vertices.  There is no accumulation nor appending of elements inside this function.

Comment (nselliott) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp]: `make_element_3d` is different because it has to handle both elements and subelements.  Each call creates a single element, but that element is made up of subelements (faces) that can be shared by two elements.  The `{i|j|k}faces` arguements are there so that the function can first figure out if a needed subelement already exists, then use it if it does, or create a new one and append it if it doesn't.

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (nselliott) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp]: Probably so.  I brute-forced this a bit when I ripped out an earlier version that used the previous layout from the Conduit 0.5.1 documentation.

File: src/libs/blueprint/conduit_blueprint_mpi_mesh.cpp
Comment (nselliott) [reply src/libs/blueprint/conduit_blueprint_mpi_mesh.cpp]: Yes, it's just my style preference to not have an if statement go beyond two lines, but I don't object to changing it to your way.

File: src/libs/blueprint/conduit_blueprint_mpi_mesh.hpp
Comment (nselliott) [reply src/libs/blueprint/conduit_blueprint_mpi_mesh.hpp]: Could be `structured_to_polygonal`?  Strictly speaking, the input mesh doesn't have to be an AMR mesh.

File: src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp
Comment (nselliott) [reply src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp]: Let's see if this works.  2 domains of differing resolutions, and the coarse elements at the interface will get extra vertices in the transformation.

```
_______________________
|    |    |__|__|__|__|
|____|____|__|__|__|__|
|    |    |__|__|__|__|
|____|____|__|__|__|__|
```

File: src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp
Comment (nselliott) [reply src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp]: Yes, this is true.

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (nselliott) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp]: `ref_win` comes from the `adjset` and defines a 1d boundary interface with an adjacent domain, but it doesn't contain information about which side of the domain (left, right, lower, upper) that it's on, so the function figures it out by comparing it to information from the domain's topology.  It figures out whether ref_win is a i face or a j face based on which direction is width 1, and then it compares its origin coordinates to `i_lo` or `j_lo` to figure out if it is on the lower or upper side.  It then traverses the length of `ref_win` and calls `make_element_2d` for to create every domain element along that side of the domain.  `iwidth` is used to map each element to an offset within the domain based on the formula `offset = j*iwidth + i`.

An alternative would be to pass the topology itself, and then extract needed values inside the function, but these three values had already been extracted before the code gets to the one place that this function is called.

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (nselliott) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp]: I probably should have deleted `connect_elements_3d` as it doesn't really do anything at this point.  Connecting the 2D elements is complicated, but it's ultimately just adding new vertices to polygons.  In 3D there are new faces created, new vertices added to old faces, and then this information has to be sent to any element that touches an affected face.

File: src/libs/blueprint/CMakeLists.txt
Comment (cyrush) [inline src/libs/blueprint/CMakeLists.txt]: quick note: we don't want the base blueprint lib to depend on mpi, only the blueprint_mpi lib

File: src/libs/blueprint/conduit_blueprint_mesh.cpp
Comment (xjrc) [reply src/libs/blueprint/conduit_blueprint_mesh.cpp]: After reviewing your comments and the content of `conduit::blueprint::mesh::connectivity::create_elements_2d`, I understand how this works now. As I have alluded to elsewhere, I feel like this interface suggests that there should be some form of superstructure that encapsulates all of these bits of functionality in a more legible way (e.g. `struct`/`class` a la `ShapeTopology`).

File: src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp
Comment (xjrc) [inline src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp]: Beautiful; thank you for adding this, Noah!

File: src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp
Comment (xjrc) [inline src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp]: I think it would be good if there were a few more tests here to ensure that the new mesh has the expected structure, e.g.:

```
EXPECT_TRUE(conduit::blueprint::mesh::topology::unstructured::verify(poly["path/to/topo"]));
EXPECT_EQ(poly["path/to/topo/elements/shape"].as_string(), "polygonal");
```

Comment (xjrc) [reply src/tests/blueprint/t_blueprint_mpi_mesh_transform.cpp]: Also, some broad-level checks to ensure that the contents look correct:

```
EXPECT_EQ(/*# elements in polygonal topology*/, /*# elements in original structured AMR topology*/);
```

File: src/libs/blueprint/CMakeLists.txt
Comment (nselliott) [reply src/libs/blueprint/CMakeLists.txt]: Good catch--I temporarily had some MPI code in conduit_blueprint_mesh before conduit_blueprint_mpi_mesh existed, but this was not intended to be merged.

**Comment by cyrush** [comment]
@nselliott  @xjrc  -- what do we still have pending here, are there more changes in the works or do we think this is ready to go in?

**Comment by nselliott** [comment]
> @nselliott @xjrc -- what do we still have pending here, are there more changes in the works or do we think this is ready to go in?

The one remaining missing piece is the correct handling of the case where a coarse element receives new hanging vertices on more than one of its faces.  The new vertices on the edges shared between those faces are duplicated rather than treated as a single vertex.  This is not easy to explain in text, so I will try to make a crude PowerPoint drawing and send email.

**Comment by nselliott** [comment]
@cyrush @xjrc OK, I consider this ready to be merged.

**Comment by cyrush** [comment]
thanks @nselliott !

**Pull request merged**
